File: TorrentAdder/SYComputerCell.m
//
//  SYComputerCell.m
//  TorrentAdder
//
//  Created by rominet on 20/05/14.
//  Copyright (c) 2014 Syan. All rights reserved.
//

#import "SYComputerCell.h"
#import "SYComputerModel.h"
#import "SYNetworkManager.h"

#define COLOR_YES ([UIColor greenColor])
#define COLOR_NO  ([UIColor   redColor])

@interface SYComputerCell ()
@property (weak,   nonatomic) IBOutlet UILabel      *nameLabel;
@property (weak,   nonatomic) IBOutlet UILabel      *hostLabel;
@property (weak,   nonatomic) IBOutlet UIImageView  *statusImageView;
@property (weak,   nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;
@end

@implementation SYComputerCell

- (void)setComputer:(SYComputerModel *)computer
{
    self->_computer = computer;
    [self updateStatus];
    [self updateTexts];

    if (self.computer)
    {
        [[NSNotificationCenter defaultCenter] removeObserver:self
                                                        name:SYNetworkManagerComputerStatusChangedNotification
                                                      object:self.computer];
        
    }

    self->_computer = computer;
    
    if (self.computer)
    {
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(notificationComputerStatusChanged:)
                                                     name:SYNetworkManagerComputerStatusChangedNotification
                                                   object:self.computer];
    }
}

- (void)dealloc
{
    if (self.computer)
    {
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(notificationComputerStatusChanged:)
                                                     name:SYNetworkManagerComputerStatusChangedNotification
                                                   object:self.computer];
    }
}

- (void)setIsAvailableComputersList:(BOOL)isAvailableComputersList
{
    [self setAccessoryType:(isAvailableComputersList ?
                            UITableViewCellAccessoryDisclosureIndicator :
                            UITableViewCellAccessoryNone)];
    [self updateTexts];
    
}

- (void)updateTexts
{
    if (self.computer)
    {
        [self.nameLabel setText:self.computer.name];
        if (self.isAvailableComputersList)
        {
            [self.hostLabel setText:self.computer.host];
        }
        else
        {
            [self.hostLabel setText:[NSString stringWithFormat:@"%@:%d",
                                     self.computer.host, self.computer.port]];
        }
    }
    else
    {
        [self.nameLabel setText:@"Add a custom computer"];
        [self.hostLabel setText:@"in case yours wasn't detected"];
    }
}

- (void)updateStatus
{
    SYComputerStatus status = SYComputerStatus_Unknown;
    if (!self.isAvailableComputersList)
    {
        status = [[SYNetworkManager shared] statusForComputer:self.computer];
        SYComputerStatus previousStatus = [[SYNetworkManager shared] previousStatusForComputer:self.computer];
        if (status == SYComputerStatus_Waiting && previousStatus != SYComputerStatus_Unknown)
            status = previousStatus;
    }
    if (self.isAvailableComputersList &&  self.computer)
        status = SYComputerStatus_Opened;
    if (self.isAvailableComputersList && !self.computer)
        status = SYComputerStatus_Waiting;
    
    switch (status)
    {
        case SYComputerStatus_Unknown:
            [self.statusImageView setImage:nil];
            [self.activityIndicator stopAnimating];
            break;
        case SYComputerStatus_Waiting:
            [self.statusImageView setImage:nil];
            [self.activityIndicator startAnimating];
            break;
        case SYComputerStatus_Closed:
            [self.statusImageView setImage:[UIImage imageNamed:@"traffic_grey"]];
            [self.activityIndicator stopAnimating];
            break;
        case SYComputerStatus_Opened:
            [self.statusImageView setImage:[UIImage imageNamed:@"traffic_green"]];
            [self.activityIndicator stopAnimating];
            break;
    }
}

- (void)notificationComputerStatusChanged:(NSNotification *)notification
{
    [self updateStatus];
}

@end

File: TorrentAdder/SYComputerCell.swift
//
//  SYComputerCell.swift
//  TorrentAdder
//
//  Created by Stanislas Chevallier on 28/11/2018.
//  Copyright © 2018 Syan. All rights reserved.
//

import UIKit

class SYComputerCell: UITableViewCell {
    
    // MARK: Init
    override func awakeFromNib() {
        NotificationCenter.default.addObserver(self, selector: #selector(self.updateStatus), name: .clientStatusChanged, object: nil)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self, name: .clientStatusChanged, object: nil)
    }
    
    // MARK: Properties
    var isAvailableComputersList: Bool = false {
        didSet {
            updateContent()
            updateStatus()
        }
    }
    var computer: SYComputerModel? {
        didSet {
            updateContent()
            updateStatus()
        }
    }
    
    // MARK: Views
    @IBOutlet private var nameLabel: UILabel!
    @IBOutlet private var hostLabel: UILabel!
    @IBOutlet private var statusImageView: UIImageView!
    @IBOutlet private var activityIndicator: UIActivityIndicatorView!
    
    // MARK: Content
    private func updateContent() {
        accessoryType = isAvailableComputersList ? .disclosureIndicator : .none
        if let computer = computer {
            nameLabel.text = computer.name
            if isAvailableComputersList {
                hostLabel.text = computer.host
            }
            else {
                hostLabel.text = [computer.host ?? "", String(computer.port)].joined(separator: ":")
            }
        }
        else if isAvailableComputersList
        {
            nameLabel.text = "Add a custom computer"
            hostLabel.text = "in case yours wasn't detected"
        }
        else
        {
            nameLabel.text = nil
            hostLabel.text = nil
        }
    }
    
    @objc private func updateStatus() {
        let loading = (isAvailableComputersList && computer == nil) ? true    : SYClientStatusManager.shared.isComputerLoading(computer)
        let status  = (isAvailableComputersList && computer != nil) ? .online : SYClientStatusManager.shared.lastStatusForComputer(computer)
        
        // show loading only if previous status is unknown, else show last status
        if loading && status == .unknown {
            statusImageView.image = nil
            activityIndicator.startAnimating()
            return
        }
        
        switch status {
        case .online:
            statusImageView.image = UIImage(named: "traffic_green")
            activityIndicator.stopAnimating()
        case .offline:
            statusImageView.image = UIImage(named: "traffic_grey")
            activityIndicator.stopAnimating()
        case .unknown:
            statusImageView.image = nil
            activityIndicator.stopAnimating()
        }
    }
}
